AsyncDisplayKit/Private/ASDisplayNode+AsyncDisplay.mm

```diff
@@ -179,141 +188,119 @@ - (void)_recursivelyRasterizeSelfAndSublayersWithIsCancelledBlock:(asdisplaynode
+
+  _propertyLock.unlock();
+  
+  // Any form of display is not valuable if we can't size the graphics buffer to use.
+  if (CGRectIsEmpty(bounds)) {
+    return nil;
+  }
+  
+  ASDisplayNodeAssert(contentsScaleForDisplay != 0.0, @"Invalid contents scale");
+  ASDisplayNodeAssert(usesInstanceMethodDisplay == NO || (flags.implementsDrawRect == NO && flags.implementsImageDisplay == NO),
+                      @"Node %@ should not implement both class and instance method display or draw", self);
+  ASDisplayNodeAssert(usesImageDisplay == NO || usesDrawRect == NO,
+                      @"Node %@ should not implement both -display and -drawRect", self);
+  ASDisplayNodeAssert(rasterizing || !(_hierarchyState & ASHierarchyStateRasterized),
+                      @"Rasterized descendants should never display unless being drawn into the rasterized container.");
+
+  if (shouldBeginRasterizing == YES) {
     // Collect displayBlocks for all descendants.
     NSMutableArray *displayBlocks = [NSMutableArray array];
     [self _recursivelyRasterizeSelfAndSublayersWithIsCancelledBlock:isCancelledBlock displayBlocks:displayBlocks];
-
-    CGFloat contentsScaleForDisplay = self.contentsScaleForDisplay;
-    BOOL opaque = self.opaque && CGColorGetAlpha(self.backgroundColor.CGColor) == 1.0f;
-
-    ASDisplayNodeAssert(self.contentsScaleForDisplay != 0.0, @"Invalid contents scale");
+    CHECK_CANCELLED_AND_RETURN_NIL();
+    
+    // If [UIColor clearColor] or another semitransparent background color is used, include alpha channel when rasterizing.
+    // Unlike CALayer drawing, we include the backgroundColor as a base during rasterization.
+    CGColorRef backgroundColor = _layer.backgroundColor;
+    opaque = (opaque && backgroundColor != NULL && fabs(CGColorGetAlpha(_layer.backgroundColor) - 1.0f) < 0.01f);
 
     displayBlock = ^id{
-      __ASDisplayLayerIncrementConcurrentDisplayCount(asynchronous, rasterizing);
-      if (isCancelledBlock()) {
-        __ASDisplayLayerDecrementConcurrentDisplayCount(asynchronous, rasterizing);
-        return nil;
-      }
-
-      ASDN_DELAY_FOR_DISPLAY();
+      DISPLAY_COUNT_INCREMENT();
+      CHECK_CANCELLED_AND_RETURN_NIL();
+      
       UIGraphicsBeginImageContextWithOptions(bounds.size, opaque, contentsScaleForDisplay);
 
       for (dispatch_block_t block in displayBlocks) {
-        if (isCancelledBlock()) {
-          UIGraphicsEndImageContext();
-          __ASDisplayLayerDecrementConcurrentDisplayCount(asynchronous, rasterizing);
-          return nil;
-        }
+        CHECK_CANCELLED_AND_RETURN_NIL(UIGraphicsEndImageContext());
         block();
       }
-
+      
       UIImage *image = UIGraphicsGetImageFromCurrentImageContext();
       UIGraphicsEndImageContext();
 
-      __ASDisplayLayerDecrementConcurrentDisplayCount(asynchronous, rasterizing);
-
-      return image;
-    };
-  } else if (flags.implementsInstanceImageDisplay || flags.implementsImageDisplay) {
```

Comment:
I was able to entirely eliminate this third condition, and thus use just one else {